Separating k8s-specific and generic logic in conversion-gen

wk8 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

`conversion-gen` currently mixes both k8s-specific logic (e.g. using conversion schemes/scopes across packages, how it deals with vendored packages, etc...), and generic conversion logic, i.e. just looking for structs with the same names in two different packages and generating efficient conversion code from one to the other.

This patch is separating these two: the generic conversion logic is moved to a new `generic-conversion-gen` generator, while the k8s-specific logic stays in conversion-gen - and uses the new generic generator under the hood to do the heavy lifting of generating conversion code.

That new version of `conversion-gen` is a drop-in replacement for the existing one: the generated code is exactly the same; it's a seamless change from k8s' perspective.

This allows writing conversion generators in other repos, and will be re-used immediately in [csi-proxy](https://github.com/kubernetes-csi/csi-proxy/tree/15613db6b0fa6b2b183e1ecc52b3bfd64b4a5005/client/api#detailed-breakdown-of-generated-files).

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes-csi/csi-proxy/issues/8

**Special notes for your reviewer**:

This patch was initially proposed on gengo, but code-generator was deemed to be a better place for this (see https://github.com/kubernetes/gengo/pull/155#issuecomment-537585332 and the discussion on that PR).

The downside is that it's less clear how to write tests in code-generator, as there doesn't seem to currently be any? Any advice welcome as to how to add tests on this; maybe the changes it makes to checked-in generated files is considered test enough?

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

A bit of documentation on how the CSI-proxy repo is planning on using this: https://github.com/kubernetes-csi/csi-proxy/tree/15613db6b0fa6b2b183e1ecc52b3bfd64b4a5005/client/api#detailed-breakdown-of-generated-files
<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: https://github.com/kubernetes-csi/csi-proxy/tree/15613db6b0fa6b2b183e1ecc52b3bfd64b4a5005/client/api#detailed-breakdown-of-generated-files
-->
```docs
- [Other doc]: https://github.com/kubernetes-csi/csi-proxy/tree/15613db6b0fa6b2b183e1ecc52b3bfd64b4a5005/client/api#detailed-breakdown-of-generated-files
```

wk8 commented:
@sttts : would welcome a review, as per our discussion on https://github.com/kubernetes/gengo/pull/155. Thanks!

wk8 commented:
/test pull-kubernetes-e2e-gce

wk8 commented:
/test pull-kubernetes-integratio

wk8 commented:
/test pull-kubernetes-integration

wk8 commented:
/test pull-kubernetes-integration

wk8 commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

roycaihw commented:
/assign @lavalamp

ddebroy commented:
/assign @sttts

lavalamp commented:
This is an enormous change. Can we get where you want to go in a series of smaller changes that are easier to evaluate? It is hard to see what is going on in this one.

I have other urgent priorities for this release and I think @sttts does, too; just speaking for myself, I need my side-project PR reviews to fit into 10-20 minutes of focus time. :/

wk8 commented:
@lavalamp : it's not so big a change as it might look at a first glance. Most of the code in this PR is copy-pasted straight from the current `staging/src/k8s.io/code-generator/cmd/conversion-gen/generators/conversion.go`, and breaking it into 4 files:
* staging/src/k8s.io/code-generator/cmd/generic-conversion-gen/generators/generic_conversion.go is where the generic conversion logic lives, inside a new `ConversionGenerator` struct
* staging/src/k8s.io/code-generator/cmd/generic-conversion-gen/generators/manual_conversions_tracker.go creates a new struct to keep track of existing conversion functions (what used to be `getManualConversionFunctions`)
* staging/src/k8s.io/code-generator/cmd/generic-conversion-gen/generators/unsafe_conversion_arbitrator.go creates a new struct to decide whether `unsafe` conversions can be used between types (what used to be `equalMemoryTypes`)

But most of the code is really just copy-pasted from the current, existing code, just introducing hooks to make it generic enough to be re-used in other contexts than k8s itself.


> Can we get where you want to go in a series of smaller changes that are easier to evaluate?

I wish I had a better answer to this than "no, really thought about it, not sure how to". Apologies :(

lavalamp commented:
Can you refactor this for reuse? Copy-pasting this means we have multiple
non-obvious places to fix bugs.

On Fri, Oct 25, 2019 at 1:55 PM Jean Rougé <notifications@github.com> wrote:

> @lavalamp <https://github.com/lavalamp> : it's not so big a change as it
> might look at a first glance. Most of the code in this PR is copy-pasted
> straight from the current staging/src/
> k8s.io/code-generator/cmd/conversion-gen/generators/conversion.go, and
> breaking it into 4 files:
>
>    - staging/src/
>    k8s.io/code-generator/cmd/generic-conversion-gen/generators/generic_conversion.go
>    is where the generic conversion logic lives, inside a new
>    ConversionGenerator struct
>    - staging/src/
>    k8s.io/code-generator/cmd/generic-conversion-gen/generators/manual_conversions_tracker.go
>    creates a new struct to keep track of existing conversion functions (what
>    used to be getManualConversionFunctions)
>    - staging/src/
>    k8s.io/code-generator/cmd/generic-conversion-gen/generators/unsafe_conversion_arbitrator.go
>    creates a new struct to decide whether unsafe conversions can be used
>    between types (what used to be equalMemoryTypes)
>
> But most of the code is really just copy-pasted from the current, existing
> code, just introducing hooks to make it generic enough to be re-used in
> other contexts than k8s itself.
>
> Can we get where you want to go in a series of smaller changes that are
> easier to evaluate?
>
> I wish I had a better answer to this than "no, really thought about it,
> not sure how to". Apologies :(
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/83621?email_source=notifications&email_token=AAE6BFVTU3EVELE3O5VQH23QQNMNBA5CNFSM4I6ULVF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECJRGLA#issuecomment-546509612>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE6BFR43LPNMRXGPPGPEU3QQNMNBANCNFSM4I6ULVFQ>
> .
>

wk8 commented:
> Can you refactor this for reuse? Copy-pasting this means we have multiple non-obvious places to fix bugs.

I should have said "cut-paste" instead of "copy-paste", sorry. There's no duplicated code; that's the whole point of this PR. "Refactor[ing] this for reuse" is the whole point of this patch :)

lavalamp commented:
When I've made PRs like this in the past, I've put each non-trivial
movement in its own commit; that way it is easy for a reviewer to see that
it is just a move and blaze through the commits. Sorry, I know that's going
to be super annoying for you now that you've already made this PR :/

IOW, as structured it's O(N*N) reviewer time to find the source for each
destination, but if both are arranged together it's O(N)

On Fri, Oct 25, 2019 at 2:07 PM Jean Rougé <notifications@github.com> wrote:

> Can you refactor this for reuse? Copy-pasting this means we have multiple
> non-obvious places to fix bugs.
>
> @lavalamp <https://github.com/lavalamp> : I should have said "cut-paste"
> instead of "copy-paste", sorry. There's no duplicated code; that's the
> whole point of this PR. "Refactor[ing] this for reuse" is the whole point
> of this patch :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/83621?email_source=notifications&email_token=AAE6BFXJSH77HAAVOGU2B43QQNNZTA5CNFSM4I6ULVF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECJSG6Q#issuecomment-546513786>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE6BFVTKUWG2JU2GM2ZZBLQQNNZTANCNFSM4I6ULVFQ>
> .
>

wk8 commented:
@lavalamp : I'll try doing that...

wk8 commented:
But honestly yeah it's gonna be a real huge pain to do that _and_ still keep code functional and generating the exact same code for each commit. Would you be okay with those commits being there mostly to help reviewers, but not necessarily maintaining a functional/working state?

I'll rebase after review is done so there's no broken commits in the main branch after merging.

lavalamp commented:
Yes, it's OK if it's not functional between commits; we can squash after
review.

On Fri, Oct 25, 2019 at 2:34 PM Jean Rougé <notifications@github.com> wrote:

> But honestly yeah it's gonna be a real huge pain to do that *and* still
> keep code functional and generating the exact same code for each commit.
> Would you be okay with those commits being there mostly to help reviewers,
> but not necessarily maintaining a functional/working state?
>
> I'll rebase after review is done so there's no broken commits in the main
> branch after merging.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/83621?email_source=notifications&email_token=AAE6BFXVUGYYXKFN7YECKPDQQNRABA5CNFSM4I6ULVF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECJT5DQ#issuecomment-546520718>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE6BFR5A7ASY6G6RBZ2AZTQQNRABANCNFSM4I6ULVFQ>
> .
>

wk8 commented:
OK I'll do that then.

wk8 commented:
/retest


On Fri, Oct 25, 2019, 5:25 PM Kubernetes Prow Robot <
notifications@github.com> wrote:

> @wk8 <https://github.com/wk8>: The following tests *failed*, say /retest
> to rerun them all:
> Test name Commit Details Rerun command
> pull-kubernetes-verify 355b467
> <https://github.com/kubernetes/kubernetes/commit/355b4674154705a9bcc12a6f9bd37ebb4290b4cc>
> link
> <https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/83621/pull-kubernetes-verify/1187871721173028864> /test
> pull-kubernetes-verify
> pull-kubernetes-kubemark-e2e-gce-big 355b467
> <https://github.com/kubernetes/kubernetes/commit/355b4674154705a9bcc12a6f9bd37ebb4290b4cc>
> link
> <https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/83621/pull-kubernetes-kubemark-e2e-gce-big/1187871721105920000/> /test
> pull-kubernetes-kubemark-e2e-gce-big
> pull-kubernetes-e2e-gce-100-performance f9fc54f
> <https://github.com/kubernetes/kubernetes/commit/f9fc54f039aa08c2c853b2f16a7eaa6d2e6c20d7>
> link
> <https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/83621/pull-kubernetes-e2e-gce-100-performance/1187874501820420096/> /test
> pull-kubernetes-e2e-gce-100-performance
>
> Full PR test history
> <https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=83621>. Your
> PR dashboard <https://gubernator.k8s.io/pr/wk8>. Please help us cut down
> on flakes by linking to
> <https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests>
> an open issue
> <https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open> when
> you hit one in your PR.
>
> Instructions for interacting with me using PR comments are available here
> <https://git.k8s.io/community/contributors/guide/pull-requests.md>. If
> you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://go.k8s.io/bot-commands>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/83621?email_source=notifications&email_token=AATLGJ6PUEDK3UQ6KGR7SVDQQOE53A5CNFSM4I6ULVF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECJ3GJI#issuecomment-546550565>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AATLGJZ3RDNHFQW2XMKU2VDQQOE53ANCNFSM4I6ULVFQ>
> .
>

yujmo commented:
````````````

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

wk8 committed:
getManualConversionFunctions moved to a new struct

Namely, `ManualConversionsTracker`

Signed-off-by: Jean Rouge <rougej+github@gmail.com>

wk8 committed:
decisions for unsafe conversions delegated to a new struct

Namely, `UnsafeConversionArbitrator`

Also fixing a little bug with the previous implementation: if there was
a manual conversion from x to y, it would decline unsafe conversions
from x to y (correct) but also from y to x, even if there wasn't a
manual conversion for that direction (incorrect)

Signed-off-by: Jean Rouge <rougej+github@gmail.com>

wk8 committed:
Adding a generic conversion generator struct, ConversionGenerator

With all the files that come with it (args, main, etc...)

Most of the code in there comes directly from the existing conversion
generator, as will be obvious in the next commit

Signed-off-by: Jean Rouge <rougej+github@gmail.com>

wk8 committed:
Rewriting conversion gen using the generic conversion generator

All the generic conversion logic is delegated to `ConversionGenerator`

Everything that stays in
code-generator/cmd/conversion-gen/generators/conversion.go is specific
to k8s.

Signed-off-by: Jean Rouge <rougej+github@gmail.com>

wk8 commented:
@lavalamp : tried breaking it up in smaller commits. Hope that helps with the review. Thanks!

wk8 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

lavalamp commented on staging/src/k8s.io/code-generator/cmd/generic-conversion-gen/generators/manual_conversions_tracker.go:
```diff
@@ -0,0 +1,169 @@
+	"k8s.io/gengo/types"
+)
+
+// ManualConversionsTracker keeps track of manually defined conversion functions.
```
OK, this file / commit is a big improvement in factoring over the existing code, thanks. I am not sure about the location of this file, though-- should it really be under a cmd/ directory? I think it will be shared between multiple commands and therefore should be in some sort of neutral library?

lavalamp commented on staging/src/k8s.io/code-generator/cmd/conversion-gen/generators/conversion.go:
```diff
@@ -839,7 +733,7 @@ func (g *genConversion) doStruct(inType, outType *types.Type, sw *generator.Snip
 
 		// try a direct memory copy for any type that has exactly equivalent values
-		if g.useUnsafe.Equal(inMemberType, outMemberType) {
+		if g.unsafeConversionArbitrator != nil && g.unsafeConversionArbitrator.CanUseUnsafeConversion(inMemberType, outMemberType) {
```
Why would this ever be nil?

wk8 replied:
If unsafe conversions are disabled.

lavalamp commented on staging/src/k8s.io/code-generator/cmd/conversion-gen/generators/conversion.go:
```diff
@@ -405,29 +315,29 @@ type genConversion struct {
+		types:                      []*types.Type{},
+		explicitConversions:        []conversionPair{},
+		skippedFields:              map[*types.Type][]string{},
+		unsafeConversionArbitrator: unsafeConversionArbitrator,
```
If this can be nil, document that with a function godoc comment (looks like the function needs one anyway) and I suggest (if possible) constructing a default one if the user passes nil.

wk8 replied:
`nil` means unsafe conversions are disabled. Adding a comment.

lavalamp commented on staging/src/k8s.io/code-generator/cmd/generic-conversion-gen/generators/unsafe_conversion_arbitrator.go:
```diff
@@ -0,0 +1,147 @@
+	// notPossibleOneWay means that we've determined that the unsafe conversion from x to y
+	// is not possible, but that decision doesn't hold for conversions from y to x.
+	// That happens when there is a non-copy manual conversion function defined from x to y.
+	notPossibleOneWay unsafeConversionDecision = iota
```
Not asking for a change, but in the future I strongly recommend keeping refactors and behavior changes 100% separated; it makes it so much easier to review (and also, I think, to not break anything while making the change).

lavalamp commented on staging/src/k8s.io/code-generator/cmd/generic-conversion-gen/generators/unsafe_conversion_arbitrator.go:
```diff
@@ -0,0 +1,147 @@
+	possible
+)
+
+func NewUnsafeConversionArbitrator(manualConversionsTracker *ManualConversionsTracker) *UnsafeConversionArbitrator {
```
godoc comment?

lavalamp commented on staging/src/k8s.io/code-generator/cmd/generic-conversion-gen/generators/conversion_generator.go:
```diff
@@ -0,0 +1,797 @@
+	"k8s.io/klog"
+)
+
+// ConversionGenerator is a generic, configurable gengo generator to generate
```
I think this is too much to copy; can we factor out the specific bits into some sort of policy interface which the two generators can each implement?

(if this is a move I must be missing the corresponding deletion, sorry)

lavalamp replied:
Ah, I think the next commit removes significant pieces of the existing copy. Hm.

lavalamp commented:
I think we need a command-neutral place to put the generic stuff. I think the first commit (and the second commit with a few tweaks) are mergeable if you can find a better location for the code. Can you do that and send them as a separate PR? Especially since you changed the behavior and not just the shape of the code, I think merging this in several steps will let us make progress while being safe (sorry).

I'm still not sure how to review the last two commits here, and not sure how to suggest going about making them easier to review. I think when I make changes like that I end up with 20-30 git commits, I commit each little movement so that each commit is obviously correct. But once you've already done this, I'm not sure how to get it into that form--maybe there is some git wizardry level I've not yet attained that makes this possible :/

wk8 commented:
@lavalamp : thanks for the preliminary review. I'm going to make the 1st two commits in a separate PR, and ping you on that.

As for the rest, I'm not sure how to best go forward from there. I guess I could try & make that easier to review by cutting it up into smaller chunks: separate the generic component into its own file, originally as copy-pasted from the existing one. Then separate commits adding hooks into it.

That being said, this whole thing is really "just" a code generator. As long as it generates the same code as the current generator does, isn't that 95% of what we care about here? And honestly it's not like the existing generator code is production-grade either, and I really tried to make it better overall, so what are your main concerns here?

wk8 committed:
Merge branch 'master' into wk8/generic_conversion
